src/utransport.rs

```diff
@@ -115,24 +107,6 @@ pub trait UListener: Send + Sync {
     /// Because `on_receive()` is async you may choose to either `.await` it in the current context
     /// or spawn it onto a new task and await there to allow current context to immediately continue.
     async fn on_receive(&self, msg: UMessage);
-
-    /// Performs some action on receipt of an error.
-    ///
-    /// # Parameters
-    ///
-    /// * `err` - The error as `UStatus`
-    ///
-    /// # Note for `UListener` implementers
-    ///
-    /// `on_error()` is expected to return almost immediately. If it does not, it could potentially
-    /// block further message receipt. For long-running operations consider passing off received
-    /// error to a different async function to handle it and returning.
-    ///
-    /// # Note for `UTransport` implementers
-    ///
-    /// Because `on_error()` is async you may choose to either `.await` it in the current context
-    /// or spawn it onto a new task and await there to allow current context to immediately continue.
-    async fn on_error(&self, err: UStatus);
```

Comment:
invoke the on_message function with a UMessage representing the error. The UMessage contains all necessary context to find the corresponding request that the error is the outcome of.

Reply:
I remember this coming up as a question for this API when discussing uP-L2 a while back.

I liked your concept originally with UListener of having two explicit paths with on_receive and on_error.

Could you sketch out in this PR what it would look like doing handling of the error returned on a UListener for the four messages types when using a UTransport: Publish, Notification, Request, Response?

If I've missed it, please point it out :slightly_smiling_face:

Reply:
I don't think I can follow. Can you be a little more specific regarding what is unclear?

Reply:
> invoke the on_message function with a UMessage representing the error. The UMessage contains all necessary context to find the corresponding request that the error is the outcome of.

I guess you mean `on_receive`? To be honest, I prefer to handle the errors in another function to make the code clean, but it's just my preference.

Reply:
Yes, @evshary is correct. What I mean @sophokles73 to ask about is how errors would be identified and handled by the UListener. My preference would be to have an explicitly different path for the handling of those errors with on_error().

Could you help me by showing examples of handling the error path within the UListener with the single on_receive()?

Reply:
In the [uProtocol Error Model](https://github.com/eclipse-uprotocol/up-spec/blob/main/basics/error_model.adoc#2-service-implementation) I tried to establish the pattern that a `commstatus` that is not `UCode::OK` indicates an error and that the receiver of the message should assume that the payload contains a protobuf encoded `UStatus` with error details.

However, some existing service implementations still return a message with a `commstatus` of `UCode::OK` but then include an (error) `UStatus` in the operations standard response message. So only the client (invoker) of the operation will ultimately know how an error manifests itself in the response message.

A uEntity will therefore (still) need to know about the error model implemented by the service operations that it invokes. The RpcClient already implements the strategy outlined above and tries to map a UMessage with a `commstatus != UCode::OK` to a `ServiceInvocationError`.

For a UTransport implementation, I would suggest to create a UMessage following the recommended error model if it runs into an unresolvable error situation and needs to inform upper layers about it. In all other cases, it should simply invoke `UListener::on_receive` with whatever message it has received from the underlying messaging infrastructure.

Reply:
Hey @sophokles73 -- reading through this now. Took last Friday off.

Reply:
Thanks for the link to the spec! Nicely written.

Now that I see the spec, this makes more sense. So within a UListener rather than having two explicit flows, one for error and one for successful execution, the implementer of a UListener would instead check the commstatus if they were specifically listening for Responses as in a Service Consumer style uEntity.

Is it correct to say that as [outlined in the UAttributes spec](https://github.com/eclipse-uprotocol/up-spec/blob/main/basics/uattributes.adoc) that the commstatus only applies to Response messages?

That in the case of Notification, Publish, and Request messages that if these messages when attempting to be sent would be in some erroneous state with commstatus set to something or something not OK they simply would not be transmitted at all by the UTransport?

Therefore that a UListener which is pinged that has been registered for Notification, Publish, or Request style messages should assume that the message it received is not in error?

Reply:
A client trying to send a Request message using UTransport:.send would get back an Err(UStatus) if the message cannot be sent. The receiving side (service provider) UListener in this case would never even get the Request message. How could it, if sending has failed?

For the same reason, a UListener listening for Notifications will simply not receive any messages for send attempts that have failed on the client (sender) side ...

Reply:
> Therefore that a UListener which is pinged that has been registered for Notification, Publish, or Request style messages should assume that the message it received is not in error?

Yes, albeit it should, of course, check whether the UAttributes are consistent and that the payload can be processed ...